src/v/kafka/server/tests/alter_config_test.cc

```diff
@@ -332,12 +332,14 @@ FIXTURE_TEST(
   test_topic_describe_configs_requested_properties, alter_config_test_fixture) {
     wait_for_controller_leadership().get();
     app.controller->get_config_frontend()
-      .invoke_on_all([](cluster::config_frontend& cfg_frontend) {
-          cluster::config_update_request r{
-            .upsert = {{"enable_schema_id_validation", "compat"}}};
-          return cfg_frontend.patch(r, model::timeout_clock::now() + 1s)
-            .discard_result();
-      })
+      .invoke_on(
+        cluster::config_frontend::version_shard,
+        [](cluster::config_frontend& cfg_frontend) {
+            cluster::config_update_request r{
+              .upsert = {{"enable_schema_id_validation", "compat"}}};
+            return cfg_frontend.patch(r, model::timeout_clock::now() + 1s)
+              .discard_result();
+        })
```

Comment:
Well, I'm not sure what I was thinking when I wrote that, I seem to recall seeing this error.

Since `config_frontend::patch` goes to the effort of dispatching to the leader node, I wonder if it shouldn't also take the trouble to dispatch `do_patch` to the correct core?

```suggestion
      .local()
      .patch(
        {.upsert = {{"enable_schema_id_validation", "compat"}}},
        model::timeout_clock::now() + 1s)
      .discard_result()
```

And something like:
```diff
diff --git a/src/v/cluster/config_frontend.cc b/src/v/cluster/config_frontend.cc
index 2ef00cdde2..5d3f40376b 100644
--- a/src/v/cluster/config_frontend.cc
+++ b/src/v/cluster/config_frontend.cc
@@ -42,7 +42,12 @@ ss::future<config_frontend::patch_result> config_frontend::patch(
           .errc = errc::no_leader_controller, .version = config_version_unset};
     }
     if (leader == _self) {
-        co_return co_await do_patch(std::move(update), timeout);
+        co_return co_await container().invoke_on(
+          cluster::config_frontend::version_shard,
+          [update{std::move(update)},
+           timeout](cluster::config_frontend& cfg_frontend) mutable {
+              return cfg_frontend.do_patch(std::move(update), timeout);
+          });
     } else {
         auto res = co_await _connections.local()
                      .with_node_client<cluster::controller_client_protocol>(
diff --git a/src/v/cluster/config_frontend.h b/src/v/cluster/config_frontend.h
index ae6f1dffc4..593169430a 100644
--- a/src/v/cluster/config_frontend.h
+++ b/src/v/cluster/config_frontend.h
@@ -13,9 +13,12 @@
 
 #include "cluster/controller_stm.h"
 
+#include <seastar/core/sharded.hh>
+
 namespace cluster {
 
-class config_frontend final {
+class config_frontend final
+  : public ss::peering_sharded_service<config_frontend> {
 public:
     // Shard ID that will track the next available version and serialize
     // writes to hand out sequential versions.
diff --git a/src/v/cluster/service.cc b/src/v/cluster/service.cc
:...skipping...
diff --git a/src/v/cluster/config_frontend.cc b/src/v/cluster/config_frontend.cc
index 2ef00cdde2..5d3f40376b 100644
--- a/src/v/cluster/config_frontend.cc
+++ b/src/v/cluster/config_frontend.cc
@@ -42,7 +42,12 @@ ss::future<config_frontend::patch_result> config_frontend::patch(
           .errc = errc::no_leader_controller, .version = config_version_unset};
     }
     if (leader == _self) {
-        co_return co_await do_patch(std::move(update), timeout);
+        co_return co_await container().invoke_on(
+          cluster::config_frontend::version_shard,
+          [update{std::move(update)},
+           timeout](cluster::config_frontend& cfg_frontend) mutable {
+              return cfg_frontend.do_patch(std::move(update), timeout);
+          });
     } else {
         auto res = co_await _connections.local()
                      .with_node_client<cluster::controller_client_protocol>(
diff --git a/src/v/cluster/config_frontend.h b/src/v/cluster/config_frontend.h
index ae6f1dffc4..593169430a 100644
--- a/src/v/cluster/config_frontend.h
+++ b/src/v/cluster/config_frontend.h
@@ -13,9 +13,12 @@
 
 #include "cluster/controller_stm.h"
 
+#include <seastar/core/sharded.hh>
+
 namespace cluster {
 
-class config_frontend final {
+class config_frontend final
+  : public ss::peering_sharded_service<config_frontend> {
 public:
     // Shard ID that will track the next available version and serialize
     // writes to hand out sequential versions.
```

Then there's a few call site that can be simplified.

But it also looks like there might be a problem with `set_next_version`, via `do_bootstrap`, that doesn't really show up, since core 0 is a default in lots of places.

Reply:
Makes sense, thanks. I've made these changes now and solved the `set_next_version` part as well by making all public methods of `config_frontend` callable from any shard.